### PR TITLE
Ingress V1: apiversion negotiation and flags, support in store, parser-k8s decoupling, ingress parsing refactor

### LINKED ISSUE #788: ingress v1: add informer and store methods to list and interract with
- Add a store method to list all IngressV1 resources (which can later be used by the parser)
- Add informers to kick the sync machinery

For #590

### PR DESCRIPTION
**Pull request by mflendrich**
closes #788 

This PR does the following (each high-level bullet point roughly matching a group of commits):
- refactors `parseIngressRules` into separate functions (without side effects) and places them in `translate.go`: `fromIngressV1beta1` and `fromTCPIngressV1beta1`. Also moves `parseKnativeIngressRules` to follow suit. All of these get merged in the end using `mergeIngressRules` recently implemented in #822.
    - this comes with a tiny, potentially behavior changing, logic modification: after this PR, default backends get added to the overall set of `ingressRules` at a different stage of _"parsing"_ than they were before.
    - this change is here as a preparation for introduction of `fromIngressV1` which will be a copy-paste of `fromIngressV1beta1` with changes as described in the ingress v1 KEP.
- removes concrete k8s types (`<your-favorite-apiversion>.Ingress`) and `TCPIngress` from `package kongstate` by introducing an intermediate `utils.K8sObjectInfo`:
    - I considered using k8s `ObjectMeta` instead of implementing `K8sObjectInfo` but that would violate separation between Kong and Kubernetes data formats that I wanted to achieve in `package kongstate`.
    - this change is here to accommodate Ingress V1.
- reimplements Ingress API version negotiation:
    - to look not only at supported API groups, but also at the specific resource kind available in that API group (because `networking.k8s.io/v1` existed in k8s before the `Ingress` kind was added thereto)
    - to support `networking.k8s.io/v1` alongside the already supported versions
    - to support enabling and disabling support for specific APIs in Kong. This is for several reasons:
        - this PR does not implement Ingress V1 fully, so we can merge this PR in with ingress v1 code paths disabled
        - if Ingress V1 support is enabled in KIC, and the apiserver is >=1.19, then KIC will use a completely new code path in parser **for all preexisting Ingress resources**, posing a significant regression risk for existing users. There are two possible strategies for Ingress v1 rollout:
            - defensive: choose to leave ingress V1 APIs disabled by default
            - less defensive but with a workaround by design: make ingress V1 APIs enabled by default, but have the disabling flag readily at hand as a workaround for users whom we break
    - drive-by side effect: makes it possible to "soft-disable" extensions/v1beta1 ingress by setting its flag default to false
- implements Ingress v1 alongside v1beta1 in store.
    - Note that the distinction between v1 and both v1beta1 types is implemented differently than between `extensions` and `networking` v1beta1.
        - This is because `extensions` and `networking` v1beta1 Ingresses are fully compatible, and they reuse the same parsing logic. Ingress V1, however, will be a separate implementation of the parsing logic. I wanted to ensure type safety (between v1beta1 and v1 ingresses), which would be impossible without differentiated storage.
- additionally: adds `*.orig` to `.gitignore`. `*.orig` files originate from manually resolved Git conflicts and are easy to accidentally commit into the repo.

Notes to the reviewer:
- Consider reviewing commit by commit.
- Commits are (intended to be) self-contained. Consider not squashing when merging.

### CODE REVIEW

**Review by rainest — APPROVED** [review]

File: internal/ingress/controller/parser/parser.go [hunk 1]
```diff
@@ -26,20 +25,22 @@ import (
...
 	}
-	parsedIngress := parseIngressRules(log, ings, tcpIngresses)
+	parsedTCPIngress := fromTCPIngressV1beta1(log, tcpIngresses)
 
 	knativeIngresses, err := s.ListKnativeIngresses()
 	if err != nil {
 		log.Errorf("failed to list Knative Ingresses: %v", err)
 	}
-	parsedKnative := parseKnativeIngressRules(knativeIngresses)
+	parsedKnative := fromKnativeIngress(knativeIngresses)
```
[the remark this answers is not in the document]
Comment (mflendrich) [reply internal/ingress/controller/parser/parser.go:41]: Looking into the implementation of `cache.ListAll`, there's no extra caching provided by `ListAll`, it's just some abstraction over what our plain Ingresses `ListXXX` are doing. In other words, I see the discrepancy as preexisting technical debt.

Could you please add a GH ticket to track this bit of debt, to be resolved at a later time?

File: cli/ingress-controller/flags.go [hunk 2]
```diff
@@ -83,6 +83,11 @@ type cliConfig struct {
 	APIServerHost      string
 	KubeConfigFilePath string
 
+	// Allowed Ingress resource versions
+	AllowIngressExtensionsV1beta1 bool
+	AllowIngressNetworkingV1beta1 bool
+	AllowIngressNetworkingV1      bool
```
Comment (hbagdi) [inline cli/ingress-controller/flags.go:89]: Why are these configuration options being added?
We want the controller to automatically figure out the resource version to use and then communicate with the API-server accordingly.

File: internal/ingress/controller/parser/util/objectmeta.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// K8sObjectInfo describes a Kubernetes object.
+type K8sObjectInfo struct {
+	Name        string
+	Namespace   string
+	Annotations map[string]string
+}
```
Comment (hbagdi) [inline internal/ingress/controller/parser/util/objectmeta.go:12]: Question (not requesting changes): Why not use metav1.ObjectMeta at this point?

File: cli/ingress-controller/main.go [hunk 4]
```diff
@@ -351,18 +344,41 @@ func main() {
 		UpdateCh: updateChannel,
 	}
 
+	var preferredIngressAPIs []utils.IngressAPI
+	if cliConfig.AllowIngressNetworkingV1 {
```
Comment (hbagdi) [inline cli/ingress-controller/main.go:348]: The commits are arranged wrong. This struct member is introduced in the next commit. That threw me off. I've a question in the next commit about the configuration option itself.
Just noting here, no changes required.

File: go.mod [hunk 5]
```diff
@@ -13,6 +13,7 @@ require (
 	github.com/kong/go-kong v0.13.0
 	github.com/lithammer/dedent v1.1.0
 	github.com/mitchellh/mapstructure v1.3.1
+	github.com/pkg/errors v0.9.1
```
Comment (hbagdi) [inline go.mod:16]: Please do not introduce this dependency. We removed it purposefully in the past. Please use Go 1.13's error wrapping where required.

File: internal/ingress/store/store.go [hunk 6]
```diff
@@ -160,11 +164,30 @@ func (s Store) GetService(namespace, name string) (*apiv1.Service, error) {
...
+func (s Store) ListIngressesV1() []*networkingv1.Ingress {
+	// filter ingress rules
+	var ingresses []*networkingv1.Ingress
+	for _, item := range s.stores.IngressV1beta1.List() {
+		ing, ok := item.(*networkingv1.Ingress)
+		if !ok {
+			s.logger.Warnf("listIngressesV1: dropping object of unexpected type: %#v", item)
+			continue
+		}
+		if !s.isValidIngressClass(&ing.ObjectMeta, s.ingressClassMatching) { // TODO
```
Comment (hbagdi) [inline internal/ingress/store/store.go:177]: Can you add in a card and add it to the 0.10 milestone to track this TODO. It is very very important that we get this right in the first go.

File: internal/ingress/store/store.go [hunk 7]
```diff
@@ -160,11 +164,30 @@ func (s Store) GetService(namespace, name string) (*apiv1.Service, error) {
 	return service.(*apiv1.Service), nil
 }
 
-// ListIngresses returns the list of Ingresses
-func (s Store) ListIngresses() []*networkingv1beta1.Ingress {
+// ListIngressesV1 returns the list of Ingresses in the Ingress v1 store.
+func (s Store) ListIngressesV1() []*networkingv1.Ingress {
+	// filter ingress rules
+	var ingresses []*networkingv1.Ingress
+	for _, item := range s.stores.IngressV1beta1.List() {
```
Comment (hbagdi) [inline internal/ingress/store/store.go:171]: Is the store being used correct? Should it be `s.stores.IngressV1.List()` instead?

File: internal/ingress/controller/parser/parser.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mflendrich) [reply internal/ingress/controller/parser/parser.go:41]: AI agreed out of band: open an issue to track this. @mflendrich

File: go.mod (hunk 5 above)
Comment (mflendrich) [reply go.mod:16]: Discussed out of band: https://godoc.org/errors#As should provide equivalent functionality using a standard Go library available as of 1.13.

File: cli/ingress-controller/flags.go (hunk 2 above)
Comment (hbagdi) [reply cli/ingress-controller/flags.go:89]: Discussed out of band.
@mflendrich Let us make this opt out instead. This means these configuration options should be of the opposite polarity and default by false.

Comment (mflendrich) [reply cli/ingress-controller/flags.go:89]: Done in dd7e691

I'm not a fan of flags with "negative semantics" (that is, I prefer `--enable-X=true` instead of `--disable-X=false` by default) because the latter usually come with an extra `!` somewhere in the code (here, in `main.go`). But it's probably not worth bikeshedding in this instance.

File: internal/ingress/controller/parser/util/objectmeta.go (hunk 3 above)
Comment (mflendrich) [reply internal/ingress/controller/parser/util/objectmeta.go:12]: Discussed out of band.

As per PR description: 

> I considered using k8s ObjectMeta instead of implementing K8sObjectInfo but that would violate separation between Kong and Kubernetes data formats that I wanted to achieve in package kongstate.

File: cli/ingress-controller/main.go (hunk 4 above)
Comment (mflendrich) [reply cli/ingress-controller/main.go]: Sorry about throwing you off.
I can squash these two commits if you wish.

File: go.mod (hunk 5 above)
Comment (mflendrich) [reply go.mod]: Done. Decided not to wrap this error.

File: internal/ingress/store/store.go (hunk 7 above)
Comment (mflendrich) [reply internal/ingress/store/store.go]: 🤦. Thanks for catching this! This is a bug which I made both in the code (`Store`) and, symmetrically, in the test (`FakeStore`).

- Added an extra test (asserting that v1 ingress store is empty if we added v1beta1, and vice versa) that would've caught this.
- Also, this is a good example why tests should have minimal logic.

File: internal/ingress/store/store.go (hunk 6 above)
Comment (mflendrich) [reply internal/ingress/store/store.go]: There is one already: #790 .

I added a comment there referencing this TODO.

Also, improved the TODO message.

File: internal/ingress/controller/parser/parser.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mflendrich) [reply internal/ingress/controller/parser/parser.go:41]: #831

**Review by hbagdi — APPROVED** [review]
Did a skim review.
1. We need to work in future to further divide this into further smaller chunks.
2. My personal perference is to have clear commits rather than adding additional commits as review goes on. That makes for a clean git history.

**Comment by hbagdi** [comment]
1. Please resolve the conflict.
2. Please merge using a merge commit. Please include the body of the PR in the merge commit message (Github won't do that for you).

**Review by rainest — APPROVED** [review]

**Review by hbagdi — APPROVED** [review]
